extensions/kinesis/src/main/java/com/hazelcast/jet/kinesis/impl/KinesisSinkP.java

```diff
@@ -0,0 +1,354 @@
+    // on, but can we affort not to?
+
+    /**
+     * PutRecords requests are limited to 500 records.
+     */
+    private static final int MAX_RECORDS_IN_REQUEST = 500;
+
+    private static final long PAUSE_AFTER_FAILURE = 1000L; //todo: exponential backoff
+
+    /**
+     * Each record, when encoded as a byte array, is limited to 1M,
+     * including the partition key (Unicode String).
+     */
+    private static final int MAX_RECORD_SIZE_IN_BYTES = 1024 * 1024;
+
+    /**
+     * The maximum allowed size of all the records in a PutRecords request,
+     * including partition keys is 5M.
+     */
+    private static final int MAX_REQUEST_SIZE_IN_BYTES = 5 * 1024 * 1024;
+
+    /**
+     * The number of Unicode characters making up keys is limited to a maximum
+     * of 256.
+     */
+    private static final int MAX_UNICODE_CHARS_IN_KEY = 256;
+
+    private final AmazonKinesisAsync kinesis;
+    private final String stream;
+
+    private final Buffer<T> buffer;
+
+    private ILogger logger;
+    private KinesisHelper helper;
+
+    private State state = State.READY_TO_SEND;
+    private long nextSendTime;
+    private Future<PutRecordsResult> sendResult;
+
+    public KinesisSinkP(
+            AmazonKinesisAsync kinesis,
+            @Nonnull String stream,
+            @Nonnull FunctionEx<T, String> keyFn,
+            @Nonnull FunctionEx<T, byte[]> valueFn
+    ) {
+        this.kinesis = kinesis;
+        this.stream = stream;
+        this.buffer = new Buffer<>(keyFn, valueFn);
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return true;
+    }
+
+    @Override
+    public void init(@Nonnull Outbox outbox, @Nonnull Context context) {
+        logger = context.logger();
+        helper = new KinesisHelper(kinesis, stream, logger);
+        helper.waitForStreamToActivate();
+    }
+
+    @Override
+    public boolean tryProcessWatermark(@Nonnull Watermark watermark) {
+        return true; //watermark ignored
+    }
+
+    @Override
+    public void process(int ordinal, @Nonnull Inbox inbox) {
+        switch (state) {
+            case READY_TO_SEND:
+                handleReadyToSend(inbox);
+                return;
+            case SENDING_IN_PROGRESS:
+                handleSendingInProgress();
+                return;
+            default:
+                throw new RuntimeException("Programming error, unhandled state: " + state);
+        }
+
+    }
+
+    private void bufferFromInbox(@Nonnull Inbox inbox) {
+        if (buffer.isFull()) {
+            return;
+        }
+
+        while (true) {
+            T t = (T) inbox.peek();
+            if (t == null) {
+                //no more items in inbox
+                return;
+            }
+
+            boolean canBeBuffered = buffer.add(t);
+            if (canBeBuffered) {
+                inbox.remove();
+            } else {
+                //no more room in buffer
+                return;
+            }
+        }
```

Comment:
Code was correct, but it's a long reading ;)
```suggestion
        for (T t; (t = (T) inbox.peek()) != null && buffer.add(t); ) {
            inbox.remove();
        }
```